packages/react-sdk/src/client.ts

```diff
@@ -0,0 +1,595 @@
+  [key: string]: boolean | number | string | null
+}
+
+export type UserAttributes = { [attribute: string]: any }
+
+type EventTags = { [tagKey: string]: boolean | number | string }
+
+type DisposeFn = () => void
+
+type OnUserUpdateHandler = (userInfo: UserInfo) => void
+
+export type OnReadyResult = {
+  success: boolean
+  reason?: string
+}
+
+export interface ReactSDKClient extends optimizely.Client {
+  user: UserInfo
+
+  onReady(opts?: { timeout?: number }): Promise<any>
+  setUser(userInfo: { id: string; attributes?: { [key: string]: any } }): void
+  onUserUpdate(handler: OnUserUpdateHandler): DisposeFn
+
+  activate(
+    experimentKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): string | null
+
+  getVariation(
+    experimentKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): string | null
+
+  getFeatureVariables(
+    featureKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): VariableValuesObject
+
+  getFeatureVariableString(
+    featureKey: string,
+    variableKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): string | null
+
+  getFeatureVariableInteger(
+    featureKey: string,
+    variableKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): number | null
+
+  getFeatureVariableBoolean(
+    featureKey: string,
+    variableKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): boolean | null
+
+  getFeatureVariableDouble(
+    featureKey: string,
+    variableKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): number | null
+
+  isFeatureEnabled(
+    featureKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): boolean
+
+  getEnabledFeatures(
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): Array<string>
+
+  track(
+    eventKey: string,
+    overrideUserId?: string | EventTags,
+    overrideAttributes?: UserAttributes,
+    eventTags?: EventTags,
+  ): void
+
+  setForcedVariation(
+    experiment: string,
+    overrideUserIdOrVariationKey: string,
+    variationKey?: string | null,
+  ): boolean
+
+  getForcedVariation(experiment: string, overrideUserId?: string): string | null
+}
+
+type UserInfo = {
+  id: string | null
+  attributes: UserAttributes
+}
+
+export const DEFAULT_ON_READY_TIMEOUT = 5000
+
+class OptimizelyReactSDKClient implements ReactSDKClient {
+  public initialConfig: optimizely.Config
+  public user: UserInfo = {
+    id: null,
+    attributes: {},
+  }
+  public notificationCenter: optimizely.NotificationCenter
+
+  private userPromiseResovler: (user: UserInfo) => void
+  private userPromise: Promise<OnReadyResult>
+  private isUserPromiseResolved: boolean = false
+  private onUserUpdateHandlers: OnUserUpdateHandler[] = []
+
+  public client: optimizely.Client
```

Comment:
Should this be public? We can have a getter for it to avoid allowing the user to accidentally updating this instance.